From b1df33af055f06ab1fa8113610f8b59bbd400356 Mon Sep 17 00:00:00 2001
From: rahulr <rahulr@maxlinear.com>
Date: Tue, 19 Nov 2024 15:16:20 +0530
Subject: [PATCH] WLANRTSYS-86232 6g ssid config is not getting applied

Issue: 6g ssid config is not getting applied
Root Cause: pwhm will send UPDATE_BEACON to hostapd when ssid is
changed, but with mbssid enabled case , pwhm will always send the cmd to
wlan4 (dummy vap).
Fix: update all beacons in 6G when ssid is changed

FIXUP to WLANRTSYS-53245: Add support for MBSSID

Signed-off-by: rahulr <rahulr@maxlinear.com>
---
 hostapd/ctrl_iface.c     |  4 ++++
 hostapd/mxl_ctrl_iface.c | 32 ++++++++++++++++++++++++++++++++
 hostapd/mxl_ctrl_iface.h |  2 ++
 3 files changed, 38 insertions(+)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index cbd2a0688..e26967eaa 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -6420,7 +6420,11 @@ static int hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 		if (hostapd_ctrl_iface_disable(hapd->iface))
 			reply_len = -1;
 	} else if (os_strcmp(buf, "UPDATE_BEACON") == 0) {
+#ifndef CONFIG_VENDOR_MXL
 		if (ieee802_11_set_beacon(hapd))
+#else
+		if (mxl_hostapd_ctrl_iface_set_beacon(hapd))
+#endif /* CONFIG_VENDOR_MXL */
 			reply_len = -1;
 #ifdef CONFIG_TESTING_OPTIONS
 	} else if (os_strncmp(buf, "RADAR ", 6) == 0) {
diff --git a/hostapd/mxl_ctrl_iface.c b/hostapd/mxl_ctrl_iface.c
index 7b578a6e8..c2147a45b 100644
--- a/hostapd/mxl_ctrl_iface.c
+++ b/hostapd/mxl_ctrl_iface.c
@@ -3071,6 +3071,22 @@ static int mxl_hostapd_ctrl_iface_set_neighbor_psd(struct hostapd_data *hapd, ch
 	return 0;
 }
 
+int mxl_hostapd_ctrl_iface_update_mbss_beacon(struct hostapd_data *hapd)
+{
+	struct hostapd_data *bss;
+	int idx;
+
+	for (idx = MULTIBSS_REFERENCE_BSS_IDX; idx < hapd->iface->num_bss; idx++) {
+		bss = hapd->iface->bss[idx];
+		if (ieee802_11_set_beacon(bss)) {
+			wpa_printf(MSG_WARNING, "%s: update beacon failed",bss->conf->iface);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
 int mxl_hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 					      char *buf, char *reply,
 					      int reply_size,
@@ -3763,6 +3779,22 @@ int mxl_hostapd_handle_wmm_ac_params(struct hostapd_data *hapd, char *cmd, char
 	return 0;
 }
 
+int mxl_hostapd_ctrl_iface_set_beacon(struct hostapd_data *hapd)
+{
+	enum hostapd_iface_state istate = hapd->iface->state;
+
+	if (istate == HAPD_IFACE_ENABLED) {
+		if (hapd->iconf->multibss_enable)
+			return mxl_hostapd_ctrl_iface_update_mbss_beacon(hapd);
+		else
+			return ieee802_11_set_beacon(hapd);
+	}
+
+	wpa_printf(MSG_WARNING, "%s: set beacon in interface state %s not allowed",
+		hapd->conf->iface, hostapd_state_text(istate));
+	return -1;
+}
+
 int mxl_hostapd_ctrl_iface_acl_add_mac(struct mac_acl_entry **acl, int *num, const char *cmd)
 { return hostapd_ctrl_iface_acl_add_mac (acl, num, cmd); }
 int mxl_hostapd_ctrl_iface_acl_del_mac(struct mac_acl_entry **acl, int *num, const char *txtaddr)
diff --git a/hostapd/mxl_ctrl_iface.h b/hostapd/mxl_ctrl_iface.h
index 9cb8270f3..312485358 100644
--- a/hostapd/mxl_ctrl_iface.h
+++ b/hostapd/mxl_ctrl_iface.h
@@ -24,6 +24,8 @@ int mxl_hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 					      char *buf, char *reply,
 					      int reply_size,
 					      int *reply_len);
+int mxl_hostapd_ctrl_iface_set_beacon(struct hostapd_data *hapd);
+int mxl_hostapd_ctrl_iface_update_mbss_beacon(struct hostapd_data *hapd);
 char * mxl_dfs_info(struct hostapd_channel_data *chan);
 int mxl_hostapd_ctrl_iface_req_beacon(struct hostapd_data *hapd, char *cmd, char *reply, size_t reply_size);
 int mxl_hostapd_ctrl_iface_set(struct hostapd_data *hapd, char *cmd, int *ret);
-- 
2.43.0

